node/src/main/scala/coop/rchain/node/proposescript/ProposeLeaderful.scala

```diff
@@ -0,0 +1,351 @@
+              phloLimit = proposeInfo.phloLimit,
+              validAfterBlockNumber = validAfterBlockNumber
+            )
+            signedDeploy = Signed(
+              d,
+              Secp256k1,
+              PrivateKey(Base16.decode(proposeInfo.deployPrivateKey).get)
+            )
+            deployResp <- deployService.doDeploy(DeployData.toProto(signedDeploy))
+            _ <- Task
+                  .raiseError(
+                    new Exception(
+                      s"Deploy on $host:$grpcPort failed on ${deployResp.message.error} with $signedDeploy"
+                    )
+                  )
+                  .whenA(deployResp.message.isError || deployResp.message.result.isEmpty)
+            deployResult = deployResp.message.result.get
+          } yield deployResult
+      )
+
+  def propose(isAsync: Boolean): Task[String] =
+    proposeApi
+      .use(
+        proposeService =>
+          for {
+            proposeResp <- proposeService.propose(ProposeQuery(isAsync = isAsync))
+            _ <- Task
+                  .raiseError(
+                    new Exception(
+                      s"Propose on $host:$grpcPort failed on ${proposeResp.message.error} with isAsync = $isAsync"
+                    )
+                  )
+                  .whenA(proposeResp.message.isError || proposeResp.message.result.isEmpty)
+            proposeResult = proposeResp.message.result.get
+          } yield proposeResult
+      )
+
+  def proposeResult: Task[String] =
+    proposeApi
+      .use(
+        proposeService =>
+          for {
+            proposeResResp <- proposeService.proposeResult(ProposeResultQuery())
+            _ <- Task
+                  .raiseError(
+                    new Exception(
+                      s"Getting propose result on $host:$grpcPort failed on ${proposeResResp.message.error}"
+                    )
+                  )
+                  .whenA(proposeResResp.message.isError || proposeResResp.message.result.isEmpty)
+            proposeRes = proposeResResp.message.result.get
+          } yield proposeRes
+      )
+  def isContainBlock(blockHash: String): Task[Boolean] =
+    getBlock(blockHash = blockHash).redeem(_ => false, _ => true)
+
+  private def deployAndPropose(isAsync: Boolean): Task[String] =
+    for {
+      latestBlocks  <- getLatestBlocks(depth = 1)
+      latestBlock   = latestBlocks.head
+      _             <- deploy(latestBlock.blockNumber)
+      proposeResult <- propose(isAsync = isAsync)
+    } yield proposeResult
+
+  def deployAndProposeAsync: Task[String] =
+    for {
+      result <- deployAndPropose(isAsync = true)
+      _ <- result match {
+            case ProposeAsyncSuccessRegex(seqNum) => seqNum.pure[Task]
+            case failedResult =>
+              Task.raiseError[String](
+                new Exception(s"Propose async failed with ${failedResult} on $host: $grpcPort")
+              )
+          }
+      proposeResult <- proposeResult
```

Comment:
The async way doesn't actually make too much difference from sync way from my perspective. On this line, the result would still hang there until the result come. 
Maybe I miss something?

Reply:
The `ProposeResultQuery` is empty right now.

I was thinking that I send the `ProposeResultQuery` and the `proposeResult` can be returned **immediately** with `ProposeInProgress` or `ProposeDone`. But the query would get the result when the propose is done. Just mention this. Not sure about what is the original design goal on this.